tidb-on-kubernetes/maintain-node.md

```diff
@@ -0,0 +1,163 @@
+---
+title: 维护节点
+category: how-to
+---
+
+# 维护节点
+
+TiDB 是高可用数据库，可以在部分数据库节点下线的情况下正常运行，因此，我们可以安全地对底层 Kubernetes 节点进行停机维护。在具体操作时，针对 PD/TiKV/TiDB 实例的不同特性，我们还需要采取不同的操作策略。本文档将详细介绍如何对 Kuberentes 节点进行临时或长期的维护操作。
+
+> **注意：**
+>
+> 长期维护节点前，需要保证 Kubernetes 集群的剩余资源足够运行 TiDB 集群。
+
+## 维护 PD/TiDB 实例所在节点
+
+PD/TiDB 实例的迁移较快，可以采取主动驱逐实例到其它节点上的策略进行节点维护：
+
+1. 检查待维护节点上是否 TiKV 实例，假如存在 TiKV 实例，请参考 [维护 TiKV 实例所在节点](#维护-tikv-实例所在节点):
+    
+    ```shell
+    $ kubectl get pod --all-namespaces -o wide | grep <node-name>
+    ```
+	
+2. 使用 `kubectl cordon` 命令防止新的 Pod 调度到待维护节点上：
+
+    ```shell
+    $ kubectl cordon <node-name>
+    ```
+
+3. 使用 `kubectl drain` 命令将待维护节点上的数据库实例迁移到其它实例上：
+
+    ```shell
+    $ kubectl drain <node-name> --ignore-daemonsets --delete-local-data
+    ```
+
+    运行后，该节点上的 TiDB 实例会自动迁移到其它可用节点上，PD 实例则会在 5 分钟后触发[自动故障转移](/TODO/link-to-auto-failover)补齐节点。
+
+4. 此时，假如希望下线该 Kubernetes 节点，则可以将该节点删除：
+
+    ```shell
+    $ kubectl delete node <node-name>
+    ```
+
+    假如希望恢复 Kubernetes 节点，则需要在恢复节点后确认其健康状态：
+
+    ```shell
+    $ watch kubectl get node <node-name>
+    ```
+    
+    观察到节点进入 `Ready` 状态后，继续操作。
+
+5. 使用 `kubectl uncordon` 命令解除节点的调度限制:
+
+    ```shell
+    $ kubectl uncordon <node-name>
+    ```
+
+6. 观察 Pod 是否全部恢复正常运行
+
+    ```shell
+    $ watch kubectl get -n $namespace pod -o wide
+    # 或者
+    $ watch tkctl get all
+    ```
+
+    Pod 恢复正常运行后，操作结束。
+
+## 维护 TiKV 实例所在节点
+
+TiKV 实例迁移较慢，并且会对集群造成一定的数据迁移负载，因此在维护 TiKV 实例所在节点前，需要根据维护需求选择操作策略：假如是短期节点维护，则不迁移 TiKV 节点，在维护结束后原地恢复节点；假如是短期内不可恢复的节点维护，则需要规划 TiKV 的迁移工作。
+
+### 短期内可恢复的节点维护
+
+针对短期维护，我们可以通过调整 PD 集群的 `max-store-down-time` 配置来增大集群所允许的 TiKV 实例下线时间，在此时间内维护完毕恢复 Kubernetes 节点后，所有该节点上的 TiKV 实例会自动恢复。
+
+    ```shell
+    $ kubectl port-forward svc/${CLUSTER_NAME}-pd 2379:2379
+    $ pd-ctl -d 
+    ```
+
+调整 `max-store-donw-time` 到合理的的值后，后续的操作方式与 [维护 PD/TiDB 实例所在节点](#维护-pdtidb-实例所在节点) 相同。 
+
+### 短期内不可恢复的节点维护
+
+针对短期内不可恢复的节点维护，如节点长期下线等情形，需要使用 `pd-ctl` 主动通知 TiDB 集群下线对应的 TiKV 实例，再手动解除 TiKV 实例与该节点的绑定。
+
+1. 使用 `kubectl cordon` 命令防止新的 Pod 调度到待维护节点上：
+
+    ```shell
+    $ kubectl cordon <node-name>
+    ```
+
+2. 查看待维护节点上的 TiKV 实例:
+
+    ```shell
+	$ tkctl get -A tikv | grep <node-name>
+    ```
+
+3. 使用 `pd-ctl` 主动下线 TiKV 实例 
+
+    > **注意：**
+    > 
+    > 下线 TiKV 实例前，需要保证集群中剩余的 TiKV 实例数不少于 PD 配置中的 TiKV 数据副本数（配置项：max-replicas）。假如不符合该条件，需要先操作[扩容 TiKV](/TODO/link-to-scaling)。
+
+    ```shell
+    # 查看 TiKV 实例的 store-id
+    $ kubectl get tc ${CLUSTER_NAME} -ojson | jq '.status.tikv.stores | .[] | select ( .podName == "${POD_NAME}" ) | .id'
+
+    # 下线实例
+    $ kubectl port-forward svc/${CLUSTER_NAME}-pd 2379:2379
+    $ pd-ctl -d store delete ${ID}
+    ```
+
+4. 等待 store 状态(`state_name`)转移为 `Tombstone`：
+
+    ```shell
+    $ watch pd-ctl -d store ${ID}
+    ```
+
+5. 解除 TiKV 实例与节点本地盘的绑定：
+
+    ```shell
+    # 查询 Pod 使用的 PesistentVolumeClaim
+    $ kubectl get -n ${namespace} pod ${pod_name} -ojson | jq '.spec.volumes | .[] | select (.name == "tikv") | .persistentVolumeClaim.claimName'
```

Comment:
`jq` is not installed in Linux distribution by default. We need to ask the user to install `jq` tool first. For this command, how about using kubectl jsonpath? Here is an example:

```
# kubectl -n stability-cluster1 get pods stability-cluster1-tikv-0 -ojsonpath='{.spec.volumes[?(@.name=="tikv")].persistentVolumeClaim.claimName}{"\n"}'
tikv-stability-cluster1-tikv-0
```

Reply:
btw, I'm fine with `jq` version too. `jq` is more advanced than kubectl `-ojsonpath` and does more tasks. Maybe we can add a link to jq installation guide somewhere?

Reply:
Good catch, I prefer to guide user to install `jq`, which is more ergonomic